blingfire/src/lib.rs

```diff
@@ -1,100 +1,71 @@
 pub fn text_to_words(source: &str, destination: &mut String) -> Result<()> {
-    tokenize(
-        |raw_source, raw_source_len, raw_destination, raw_destination_capacity| unsafe {
-            text_to_words_ffi(
-                raw_source,
-                raw_source_len,
-                raw_destination,
-                raw_destination_capacity,
-            )
-        },
-        source,
-        destination,
-    )
+    tokenize(text_to_words_ffi, source, destination)
 }
 
 #[inline]
 pub fn text_to_sentences(source: &str, destination: &mut String) -> Result<()> {
-    tokenize(
-        |raw_source, raw_source_len, raw_destination, raw_destination_capacity| unsafe {
-            text_to_sentences_ffi(
-                raw_source,
-                raw_source_len,
-                raw_destination,
-                raw_destination_capacity,
-            )
-        },
-        source,
-        destination,
-    )
+    tokenize(text_to_sentences_ffi, source, destination)
 }
 
+type Tokenizer = unsafe extern "C" fn(*const c_char, c_int, *mut c_char, c_int) -> c_int;
+
+// The maximum length is when the destination length can't fit in the maximum blingfire array. Worst
+// case, the source is tokenized by adding a space after every character + a null character at the
+// end.
+//
+// The maximum array size is defined in blingfire-sys/BlingFire/blingfireclient.library/inc/FALimits.h
+const MAX_SOURCE_LENGTH: usize = 1_000_000_000 / 2 - 2;
+
 #[inline]
-fn tokenize<Tokenizer>(tokenizer: Tokenizer, source: &str, destination: &mut String) -> Result<()>
+fn tokenize(tokenizer: Tokenizer, source: &str, destination: &mut String) -> Result<()>
 where
-    Tokenizer: Fn(
-        *const std::os::raw::c_char,
-        std::os::raw::c_int,
-        *mut std::os::raw::c_char,
-        std::os::raw::c_int,
-    ) -> std::os::raw::c_int,
 {
     destination.clear();
 
     if source.is_empty() {
         return Ok(());
     }
 
+    let source_len = source.len();
+    ensure!(source_len <= MAX_SOURCE_LENGTH, ErrorKind::SourceTooLarge);
+    let source_len = source_len as c_int;
+
     loop {
-        let length = tokenizer(
-            source.as_ptr() as *const i8,
-            source
-                .len()
-                .try_into()
-                .chain_err(|| ErrorKind::SourceTooLarge)?,
-            destination.as_mut_ptr() as *mut i8,
-            destination
-                .capacity()
-                .try_into()
-                .chain_err(|| ErrorKind::DestinationTooLarge)?,
-        );
-
-        if length < 0 {
-            // The C++ function returned -1, an unknown error.
-            bail!(ErrorKind::UnknownError);
-        } else if length as usize > destination.capacity() {
+        let length = unsafe {
+            tokenizer(
+                source.as_ptr() as *const c_char,
+                source_len,
+                destination.as_mut_ptr() as *mut c_char,
+                destination.capacity().try_into().unwrap_or(i32::MAX),
```

Comment:
So this can technically go in an infinite loop now, as `destination.capacity()` is at most `i32::MAX` and if that's not enough for the parsed string, it will call `destination.reserve()` and call the function again (forever)

Reply:
the return type is i32, so it can't be larger i32::MAX